modules/caddyhttp/reverseproxy/selectionpolicies.go

```diff
@@ -380,6 +381,48 @@ func (s *HeaderHashSelection) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
 	return nil
 }
 
+// CookieHashSelection is a policy that selects
+// a host based on a given cookie name.
+type CookieHashSelection struct {
+	// The HTTP cookie name whose value is to be hashed and used for upstream selection.
+	Field string `json:"field,omitempty"`
+}
+
+// CaddyModule returns the Caddy module information.
+func (CookieHashSelection) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID:  "http.reverse_proxy.selection_policies.cookie",
+		New: func() caddy.Module { return new(CookieHashSelection) },
+	}
+}
+
+// Select returns an available host, if any.
+func (s CookieHashSelection) Select(pool UpstreamPool, req *http.Request, w http.ResponseWriter) *Upstream {
+	if s.Field == "" {
+		return nil
+	}
+	cookie, err := req.Cookie(s.Field)
+	var cookieValue string
+	if err != nil || cookie == nil {
+		cookieValue = caddy.RandomString(16)
+		http.SetCookie(w, &http.Cookie{Name: s.Field, Value: cookieValue, Secure: false})
+	} else {
+		cookieValue = cookie.Value
+	}
+	return hostByHashing(pool, cookieValue)
+}
```

Comment:
I think how it should work is more like this:

- Check if there's a cookie.
- If there's no cookie:
  - Get a random host (see RandomChoiceSelection)
  - Hash (maybe HMAC with some key for privacy), or encrypt (with some key, for privacy + reversibility) the upstream.Dial string as the cookie value, write the cookie.
- If there's a cookie:
  - If you chose to hash, then loop over the upstreams until you find a match.
  - If you chose to encrypt, then decrypt and lookup the upstream.
  - If there's no match, then give a random upstream again (i.e. the upstream is no longer available), write the cookie.
- Return the upstream.